src/chainstate/burn/db/sortdb.rs

```diff
@@ -1285,53 +1383,104 @@ impl <'a> SortitionDBConn <'a> {
     }
 }
 
-pub struct PoxDB;
-
-impl PoxDB {
-    /// Get the canonical PoX identifier for a given burnchain header hash.
-    ///   this result may change if a previously unknown PoX anchor is processed.
-    pub fn get_canonical_pox_id(&self, _burnchain_header_hash: &BurnchainHeaderHash) -> Result<PoxId, ()> {
-        Ok(PoxId::stubbed())
+// High-level functions used by ChainsCoordinator
+impl SortitionDB {
+    pub fn get_sortition_id(&self, burnchain_header_hash: &BurnchainHeaderHash, sortition_tip: &SortitionId) -> Result<Option<SortitionId>, BurnchainError> {
+        let handle = self.index_handle(sortition_tip);
+        handle.get_sortition_id_for_bhh(burnchain_header_hash)
+            .map_err(BurnchainError::from)
     }
-    /// Get the parent PoX identifier for a given identifier.
-    ///  If the PoX identifier does not have a parent (should only be true for the "base" PoX identifier)
-    ///    return an error.
-    pub fn get_parent_pox_id(&self, _pox_id: &PoxId) -> Result<PoxId, ()> {
-        Err(())
+
+    pub fn is_sortition_processed(&self, burnchain_header_hash: &BurnchainHeaderHash, sortition_tip: &SortitionId) -> Result<bool, BurnchainError> {
+        self.get_sortition_id(burnchain_header_hash, sortition_tip)
+            .map(|x| x.is_some())
     }
-    /// does the given pox identifier describe a child fork of the parent pox identifier? this returns true if they
-    ///   are equal or child is a descendant.
-    pub fn is_pox_id_descendant(&self, parent: &PoxId, child: &PoxId) -> bool {
-        return parent == child || self.get_parent_pox_id(child).as_ref() == Ok(parent)
+
+    /// Is the given block an expected PoX anchor in this sortition history?
+    ///  if so, return the Stacks block hash
+    pub fn is_stacks_block_pox_anchor(&mut self, block: &BlockHeaderHash, sortition_tip: &SortitionId) -> Result<Option<BlockHeaderHash>, BurnchainError> {
+        let handle = self.index_handle(sortition_tip);
+        let expects_block_as_anchor = handle.get_tip_indexed(&db_keys::pox_anchor_to_prepare_end(block))?
+            .map(|_| block.clone());
+
+        return Ok(expects_block_as_anchor)
     }
-    pub fn stubbed() -> PoxDB {
-        PoxDB
+
+    pub fn invalidate_descendants_of(&mut self, burn_block: &BurnchainHeaderHash) -> Result<(), BurnchainError> {
+        let db_tx = self.conn.transaction()?;
+        let mut queue = vec![burn_block.clone()];
```

Comment:
You may want a `VecDeque` here -- it's more efficient for implementing a queue.

Reply:
I think it's probably just a poorly named variable: it gets used as a stack (because the order of the invalidations doesn't matter), which a Vec implements efficiently.